sycl/include/sycl/detail/accessor_iterator.hpp

```diff
@@ -0,0 +1,439 @@
+    return *this;
+  }
+
+  difference_type operator-(const __accessor_iterator_base &_Rhs) {
+    if (_Rhs._MLinearCurrent > _MLinearCurrent)
+      return _Rhs._MLinearCurrent - _MLinearCurrent;
+    else
+      return _MLinearCurrent - _Rhs._MLinearCurrent;
+  }
+
+  bool operator<(const __accessor_iterator_base<_Dimensions> &_Other) const {
+    return _MLinearCurrent < _Other._MLinearCurrent;
+  }
+
+  bool operator>(const __accessor_iterator_base<_Dimensions> &_Other) const {
+    return _Other < *this;
+  }
+
+  bool operator<=(const __accessor_iterator_base<_Dimensions> &_Other) const {
+    return !(*this > _Other);
+  }
+
+  bool operator>=(const __accessor_iterator_base<_Dimensions> &_Other) const {
+    return !(*this < _Other);
+  }
+
+  bool operator==(const __accessor_iterator_base<_Dimensions> &_Other) const {
+    return _MLinearCurrent == _Other._MLinearCurrent;
+  }
+
+  bool operator!=(const __accessor_iterator_base<_Dimensions> &_Other) const {
+    return !(*this == _Other);
+  }
+
+private:
+  void __increment() {
+    if (_MLinearCurrent >= _MLinearEnd)
+      return;
+
+    ++_MLinearCurrent;
+    if (_MCurrent[_Index0] < _MEnd[_Index0])
+      _MCurrent[_Index0]++;
+    if constexpr (_Dimensions > 1) {
+      if (_MCurrent[_Index0] == _MEnd[_Index0]) {
+        if (_MCurrent[_Index1] < _MEnd[_Index1]) {
+          _MCurrent[_Index1]++;
+          _MCurrent[_Index0] = _MBegin[_Index0];
+        }
+      }
+    }
+    if constexpr (_Dimensions > 2) {
+      if (_MCurrent[_Index1] == _MEnd[_Index1]) {
+        if (_MCurrent[_Index2] < _MEnd[_Index2]) {
+          _MCurrent[_Index2]++;
+          _MCurrent[_Index0] = _MBegin[_Index0];
+          _MCurrent[_Index1] = _MBegin[_Index1];
+        }
+      }
+    }
+  }
+
+  void __decrement() {
+    if (_MLinearCurrent == _LinearBegin)
+      return;
+
+    --_MLinearCurrent;
+    if (_MCurrent[_Index0] > 0)
+      _MCurrent[_Index0]--;
+    if constexpr (_Dimensions > 1) {
+      if (_MCurrent[_Index0] == 0) {
+        if (_MCurrent[_Index1] > 0) {
+          _MCurrent[_Index1]--;
+          _MCurrent[_Index0] = _MEnd[_Index0] - 1;
+        }
+      }
+    }
+    if constexpr (_Dimensions > 2) {
+      if (_MCurrent[_Index1] == 0) {
+        if (_MCurrent[_Index2] > 0) {
+          _MCurrent[_Index2]--;
+          _MCurrent[_Index0] = _MEnd[_Index0] - 1;
+          _MCurrent[_Index1] = _MEnd[_Index1] - 1;
+        }
+      }
+    }
+  }
+
+  void __adjustForward(difference_type _N) {
+    if (_MLinearCurrent + _N > _MLinearEnd)
+      _MLinearCurrent = _MLinearEnd;
+    else
+      _MLinearCurrent += _N;
+    _MCurrent = __delinearizeIndex(_MLinearCurrent);
+  }
+
+  void __adjustBackwards(difference_type _N) {
+    if (_N > _MLinearCurrent)
+      _MLinearCurrent = _LinearBegin;
+    else
+      _MLinearCurrent -= _N;
+    _MCurrent = __delinearizeIndex(_MLinearCurrent);
+  }
+
+  size_t __linearizeIndex(const id<_Dimensions> &_Id) const {
+    size_t _Result = _Id[_Index0];
```

Comment:
Another place to use compile-time `dim_loop`. Same below.

Reply:
This is not applicable anymore since ba4557784b2781a6b750b3489e1d6cf850b4a615